### ISSUE TITLE
[JavaScript] Implement new Concept Exercise: type-checking

### ISSUE THREAD

**Issue by SleeplessByte** [exercism/javascript#946]
This issue describes how to implement the `type-checking` concept exercise for the `javascript` track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

See the documentation _above_ (general documentation), as well as [How to implement a Concept Exercise in JavaScript](https://github.com/exercism/v3/blob/master/languages/javascript/reference/implementing-a-concept-exercise.md).

## Goal

The goal of this exercise is to teach the student how they can type-check their values **at run-time**. 

## JavaScript specificness

ℹ Many languages have a way to check a value against a type `.is_a?` `instance of` etc. JavaScript is special:

- `typeof` only work for primitives
- `typeof null` is `object` 
- `instanceof` only works for _prototype_ inheritance
- `Array.isArray` is not the same as `instanceof Array`
- `'property' in object`
- `object.hasOwnProperty('property')`
- duck typing (array like, partial shapes, etc.)

## Learning objectives

- How to use `typeof` for primitives (e.g. implement `isNumeric`,`isString`)
- How to create a sane type-checking function:
  - `isNumber` should check if the value is finite (ie not `NaN`)
  - `isObject` should check if the value is actually an object (so not `null`)
  - `isNumericString` should check if the value is a string but only consists of numbers
  - etc.
- How to check for emptiness:
  - `isEmptyArray` should check if the value is an array and empty
  - `isNonEmptyArray` should check if the value is an array and not empty
  - etc.

You should have the student create a _large number of very simple functions_. So the learning objectives are all around the same object (usage of `typeof`, `instanceof` and property checking), but we want to give the student something they can use in real life. 

**Consider also having a few `assertXXX` variants**: These use the check above (`boolean` result) and return _self_ if valid, or throw an `Error` when invalid.

Take your inspiration from [this library](https://github.com/sindresorhus/is). Please add to references if you end up using the code to create your implementation.

## Out of scope

- Boxed types (use `after.md` to teach type checking boxed primitives, e.g. `isString(val) => typeof val === 'string or val instanceof String`)
- JSON Schema or similar

## Concepts

- `type-checking`

## Prerequisites

- `basics`
- ... (do you think something is missing?)

### After

Boxed types, common mistakes, and _why_ `typeof null` returns `object`.

## Help

You can choose to do this solo-style, or collaborate with multiple people on this. My suggestion would be is to 

1. **first accept this issue by saying "I'd like to work on this"** (you will not get a response from me. Just go with it) and optionally request that someone works with you (and wait for a second person to accept your request).
2. Use this issue to write down what you need (collect reference material) and discuss, if necessary, what to write and what not to write.
  - [`typeof` reference document](https://github.com/exercism/v3/blob/master/languages/javascript/reference/keywords/typeof.md)
  - [`instanceof` reference document](https://github.com/exercism/v3/blob/master/languages/javascript/reference/keywords/instanceof.md)
  - [`'in' vs .hasOwnProperty`](http://adripofjavascript.com/blog/drips/the-uses-of-in-vs-hasownproperty.html)

3. Draft a PR and request the feedback from at least one other JavaScript member (I recommend just pinging the channel on slack).
